Error on compiling.  2 variables not declared.  Arduino 1.8.3

Arduino_PID_for_Espresso_v2_0:87: error: 'menuUsed' was not declared in this scope

 MenuBackend menu = MenuBackend(menuUsed,menuChanged);

                                ^

Arduino_PID_for_Espresso_v2_0:87: error: 'menuChanged' was not declared in this scope

 MenuBackend menu = MenuBackend(menuUsed,menuChanged);

@ML-Weiss 
I changed what you said above but now we are getting these errors :S


> C:\Users\UK\AppData\Local\Temp\cc1UkDbY.ltrans2.ltrans.o: In function `__base_ctor ':
>  
> C:\Program Files (x86)\Arduino\libraries\MenuBackend/MenuBackend.h:151: undefined reference to `menuChangeEvent(MenuChangeEvent)'
>  
> C:\Program Files (x86)\Arduino\libraries\MenuBackend/MenuBackend.h:151: undefined reference to `menuChangeEvent(MenuChangeEvent)'
>  
> C:\Program Files (x86)\Arduino\libraries\MenuBackend/MenuBackend.h:152: undefined reference to `menuUseEvent(MenuUseEvent)'
>  
> C:\Program Files (x86)\Arduino\libraries\MenuBackend/MenuBackend.h:152: undefined reference to `menuUseEvent(MenuUseEvent)'
>  
> collect2.exe: error: ld returned 1 exit status
>  
> exit status 1
> Error compiling for board Arduino/Genuino Uno.
> 
The rows in the audrino is like this 
```
static void menuChangeEvent(MenuChangeEvent changed);
static void menuUseEvent(MenuUseEvent used);

//Menu variables
MenuBackend menu = MenuBackend(menuUseEvent,menuChangeEvent);
```

I found that using function prototypes fixed that problem.

On Nov 23, 2017 2:51 PM, "Christoffer Karlsson" <notifications@github.com>
wrote:

> @ML-Weiss <https://github.com/ml-weiss>
> I changed what you said above but now we are getting these errors :S
>
> C:\Users\UK\AppData\Local\Temp\cc1UkDbY.ltrans2.ltrans.o: In function
> `__base_ctor ':
>
> C:\Program Files (x86)\Arduino\libraries\MenuBackend/MenuBackend.h:151:
> undefined reference to `menuChangeEvent(MenuChangeEvent)'
>
> C:\Program Files (x86)\Arduino\libraries\MenuBackend/MenuBackend.h:151:
> undefined reference to `menuChangeEvent(MenuChangeEvent)'
>
> C:\Program Files (x86)\Arduino\libraries\MenuBackend/MenuBackend.h:152:
> undefined reference to `menuUseEvent(MenuUseEvent)'
>
> C:\Program Files (x86)\Arduino\libraries\MenuBackend/MenuBackend.h:152:
> undefined reference to `menuUseEvent(MenuUseEvent)'
>
> collect2.exe: error: ld returned 1 exit status
>
> exit status 1
> Error compiling for board Arduino/Genuino Uno.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/CoffeeTronics/Arduino-PID-Controller-for-Espresso-V2.0/issues/2#issuecomment-346701686>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABXR4JRo5a84lkZad9a06NIIuPpi_P1lks5s5ejLgaJpZM4QdxOe>
> .
>

--- a/Arduino_PID_for_Espresso_v2_0/Arduino_PID_for_Espresso_v2_0.ino
+++ b/Arduino_PID_for_Espresso_v2_0/Arduino_PID_for_Espresso_v2_0.ino
@@ -83,6 +83,11 @@ long lastLeftDebounceTime = 0; // the last time the output pin was toggled
 long lastRightDebounceTime = 0; // the last time the output pin was toggled
 long debounceDelay = 300; // the debounce time
 
+//it's necessary to manually add the prototypes for the menuUsed and menuChanged functions here because the Arduino IDE's
+//automated function prototype generation system inserts them after they are referenced in the MenuBackend instantiation
+void menuUsed(MenuUseEvent used);
+void menuChanged(MenuChangeEvent changed);
+
 //Menu variables
 MenuBackend menu = MenuBackend(menuUsed,menuChanged);
 //initialize menu items

